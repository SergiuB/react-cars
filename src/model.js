import PropTypes from 'prop-types';

const model = {
  name: {
    type: 'string',
  },
  acceleration: {
    type: 'number',
    min: 1,
    max: 100,
    step: 0.1,
  },
  cylinders: {
    type: 'number',
    min: 2,
    max: 16,
    step: 1,
  },
  displacement: {
    type: 'number',
    min: 1,
    max: 1000,
    step: 1,
  },
  horsepower: {
    type: 'number',
    min: 40,
    max: 1000,
    step: 1,
  },
  mpg: {
    type: 'number',
    min: 0,
    max: 100,
    step: 1,
  },
  weight: {
    type: 'number',
    min: 500,
    max: 10000,
    step: 1,
  },
  year: {
    type: 'number',
    min: 0,
    max: 99,
    step: 1,
  },
};


const carShape = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  acceleration: PropTypes.number,
  cylinders: PropTypes.number,
  displacement: PropTypes.number,
  horsepower: PropTypes.number,
  mpg: PropTypes.number,
  weight: PropTypes.number,
  year: PropTypes.number,
});

const fieldShape = PropTypes.shape({
  type: PropTypes.string.isRequired,
  min: PropTypes.number,
  max: PropTypes.number,
  step: PropTypes.number,
});

const modelShape = PropTypes.shape({
  name: fieldShape,
  acceleration: fieldShape,
  cylinders: fieldShape,
  displacement: fieldShape,
  horsepower: fieldShape,
  mpg: fieldShape,
  weight: fieldShape,
  year: fieldShape,
});

export {
  carShape,
  modelShape,
};

export default model;
