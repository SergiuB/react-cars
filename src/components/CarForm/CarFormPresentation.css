.car-form {
  display: flex;
  flex-direction: column;
  align-items: center;
  border: 1px;
}

.car-form-fields {
  display: flex;
  flex-direction: column;
  align-items: flex-end;
  margin: 10px;
}

.car-form-field {
  display: flex;
  flex-direction: column;
  align-items: flex-end;
}

.car-form-field .label-text {
  margin-right: 5px;
}

.car-form-buttons button {
  margin: 5px;
}