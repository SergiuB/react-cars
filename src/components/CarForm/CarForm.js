import React, { Component } from 'react';
import PropTypes from 'prop-types';
import CarFormPresentation from './CarFormPresentation';

class CarForm extends Component {
  constructor(props) {
    super(props);
    this.state = { car: { ...this.props.car } };
  }

  handleFieldChange = (fieldName, fieldValue) => {
    const fieldDef = this.props.model[fieldName];
    const fieldError = this.props.validateField(fieldValue, fieldDef);
    this.setState({
      car: {
        ...this.state.car,
        [fieldName]: fieldValue,
      },
      errors: {
        ...this.state.errors,
        [fieldName]: fieldError
      }
    });
  }

  submit = (event) => {
    this.props.onSubmit(this.state.car);
    event.preventDefault();
  }

  render() {
    const { model, onCancel } = this.props;
    return (
      <CarFormPresentation
        {...this.state}
        model={model}
        onChangeField={this.handleFieldChange}
        onSubmit={this.submit}
        onCancel={onCancel}
      />
    );
  }
}

CarForm.propTypes = {
  car: PropTypes.object.isRequired,
  model: PropTypes.object.isRequired,
  validateField: PropTypes.func.isRequired,
  onChangeField: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

CarForm.defaultProps = {
  validateField: () => '',
  onChangeField: () => {},
  onSubmit: () => {},
  onCancel: () => {}
};

export default CarForm;
