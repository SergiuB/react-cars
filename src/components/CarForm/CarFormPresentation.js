import React from 'react';
import PropTypes from 'prop-types';
import './CarFormPresentation.scss';

const CarFormPresentation = ({
    name,
    nameError,
    onChangeName,
    acceleration,
    accelerationError,
    onChangeAcceleration,
    onSubmit,
    onCancel,
  }) => {
    const hasError = !!(nameError || accelerationError);
    return (
      <form onSubmit={onSubmit}>
        <input type="text" name="name" onChange={onChangeName} value={name}/>
          { nameError && <span className="error">{ nameError }</span> }
        <input type="number" name="acceleration" onChange={onChangeAcceleration} value={acceleration}/>
          { accelerationError && <span className="error">{ accelerationError }</span> }
        <button type="submit" name="submit" disabled={hasError}>Save</button>
        <button type="button" name="cancel" onClick={onCancel}>Cancel</button>
      </form>
    );
  }

CarFormPresentation.propTypes = {
  name: PropTypes.string.isRequired,
  nameError: PropTypes.string.isRequired,
  acceleration: PropTypes.number.isRequired,
  accelerationError: PropTypes.string.isRequired,
  onChangeName: PropTypes.func.isRequired,
  onChangeAcceleration: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired,
  onCancel: PropTypes.func.isRequired,
};

CarFormPresentation.defaultProps = {
  nameError: '',
  accelerationError: '',
}

export default CarFormPresentation;