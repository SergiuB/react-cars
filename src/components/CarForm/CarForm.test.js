import React from 'react';
import { shallow } from 'enzyme';

import CarForm from './CarForm';
import model from '../../model';
import validateField from '../../validation';

const carData = {
  id: 1,
  name: 'Audi',
  acceleration: 12,
};

describe('CarForm', function () {
  it('renders correctly', function () {
    const wrapper = shallow(
      <CarForm car={carData} model={model} validateField={validateField} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('updates props in response to change handlers', function () {
    const wrapper = shallow(
      <CarForm car={carData} model={model} validateField={validateField} />
    );

    wrapper.prop('onChangeField')('acceleration', 3);
    expect(wrapper.prop('car').acceleration).toEqual(3);

    wrapper.prop('onChangeField')('name', 'VW');
    expect(wrapper.prop('car').name).toEqual('VW');
  });

  it('updates error props in response to setting invalid values', function () {
    const wrapper = shallow(
      <CarForm car={carData} model={model} validateField={validateField} />
    );

    wrapper.prop('onChangeField')('acceleration', NaN);
    expect(wrapper.prop('errors').acceleration).toEqual('Invalid number');

    wrapper.prop('onChangeField')('mpg', -100);
    expect(wrapper.prop('errors').mpg).toEqual('Value must be greater than 0');

    wrapper.prop('onChangeField')('horsepower', 100000);
    expect(wrapper.prop('errors').horsepower).toEqual('Value must be lower than or equal to 1000');

    wrapper.prop('onChangeField')('name', '');
    expect(wrapper.prop('errors').name).toEqual('Value is mandatory');
  });

  it('calls onSubmit with the changed values', function () {
    const onSubmit = jest.fn();
    const wrapper = shallow(
      <CarForm car={carData} model={model} validateField={validateField} onSubmit={onSubmit} />
    );

    wrapper.prop('onChangeField')('acceleration', 10);
    wrapper.prop('onChangeField')('name', 'VW');

    wrapper.prop('onSubmit')({ preventDefault() {} });

    expect(onSubmit).toBeCalledWith({ id: 1, acceleration: 10, name: 'VW' });
  });

  it('calls onCancel', function () {
    const onCancel = jest.fn();
    const wrapper = shallow(
      <CarForm car={carData} model={model} validateField={validateField} onCancel={onCancel} />
    );

    wrapper.prop('onCancel')();
    expect(onCancel).toBeCalled();
  });
});
