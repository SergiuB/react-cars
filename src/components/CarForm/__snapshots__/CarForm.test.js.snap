// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarForm renders correctly 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <CarFormPresentation
    car={
        Object {
            "acceleration": 12,
            "id": 1,
            "name": "Audi",
          }
    }
    errors={Object {}}
    model={
        Object {
            "acceleration": Object {
              "max": 100,
              "min": 1,
              "step": 0.1,
              "type": "number",
            },
            "cylinders": Object {
              "max": 16,
              "min": 2,
              "step": 1,
              "type": "number",
            },
            "displacement": Object {
              "max": 1000,
              "min": 1,
              "step": 1,
              "type": "number",
            },
            "horsepower": Object {
              "max": 1000,
              "min": 40,
              "step": 1,
              "type": "number",
            },
            "mpg": Object {
              "max": 100,
              "min": 0,
              "step": 1,
              "type": "number",
            },
            "name": Object {
              "type": "string",
            },
            "weight": Object {
              "max": 10000,
              "min": 500,
              "step": 1,
              "type": "number",
            },
            "year": Object {
              "max": 99,
              "min": 0,
              "step": 1,
              "type": "number",
            },
          }
    }
    onCancel={[Function]}
    onChangeField={[Function]}
    onSubmit={[Function]}
/>,
  "nodes": Array [
    <CarFormPresentation
      car={
            Object {
                  "acceleration": 12,
                  "id": 1,
                  "name": "Audi",
                }
      }
      errors={Object {}}
      model={
            Object {
                  "acceleration": Object {
                    "max": 100,
                    "min": 1,
                    "step": 0.1,
                    "type": "number",
                  },
                  "cylinders": Object {
                    "max": 16,
                    "min": 2,
                    "step": 1,
                    "type": "number",
                  },
                  "displacement": Object {
                    "max": 1000,
                    "min": 1,
                    "step": 1,
                    "type": "number",
                  },
                  "horsepower": Object {
                    "max": 1000,
                    "min": 40,
                    "step": 1,
                    "type": "number",
                  },
                  "mpg": Object {
                    "max": 100,
                    "min": 0,
                    "step": 1,
                    "type": "number",
                  },
                  "name": Object {
                    "type": "string",
                  },
                  "weight": Object {
                    "max": 10000,
                    "min": 500,
                    "step": 1,
                    "type": "number",
                  },
                  "year": Object {
                    "max": 99,
                    "min": 0,
                    "step": 1,
                    "type": "number",
                  },
                }
      }
      onCancel={[Function]}
      onChangeField={[Function]}
      onSubmit={[Function]}
/>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <CarForm
        car={
                Object {
                        "acceleration": 12,
                        "id": 1,
                        "name": "Audi",
                      }
        }
        model={
                Object {
                        "acceleration": Object {
                          "max": 100,
                          "min": 1,
                          "step": 0.1,
                          "type": "number",
                        },
                        "cylinders": Object {
                          "max": 16,
                          "min": 2,
                          "step": 1,
                          "type": "number",
                        },
                        "displacement": Object {
                          "max": 1000,
                          "min": 1,
                          "step": 1,
                          "type": "number",
                        },
                        "horsepower": Object {
                          "max": 1000,
                          "min": 40,
                          "step": 1,
                          "type": "number",
                        },
                        "mpg": Object {
                          "max": 100,
                          "min": 0,
                          "step": 1,
                          "type": "number",
                        },
                        "name": Object {
                          "type": "string",
                        },
                        "weight": Object {
                          "max": 10000,
                          "min": 500,
                          "step": 1,
                          "type": "number",
                        },
                        "year": Object {
                          "max": 99,
                          "min": 0,
                          "step": 1,
                          "type": "number",
                        },
                      }
        }
        onCancel={[Function]}
        onChangeField={[Function]}
        onSubmit={[Function]}
        validateField={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": CarForm {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "handleFieldChange": [Function],
        "props": Object {
          "car": Object {
            "acceleration": 12,
            "id": 1,
            "name": "Audi",
          },
          "model": Object {
            "acceleration": Object {
              "max": 100,
              "min": 1,
              "step": 0.1,
              "type": "number",
            },
            "cylinders": Object {
              "max": 16,
              "min": 2,
              "step": 1,
              "type": "number",
            },
            "displacement": Object {
              "max": 1000,
              "min": 1,
              "step": 1,
              "type": "number",
            },
            "horsepower": Object {
              "max": 1000,
              "min": 40,
              "step": 1,
              "type": "number",
            },
            "mpg": Object {
              "max": 100,
              "min": 0,
              "step": 1,
              "type": "number",
            },
            "name": Object {
              "type": "string",
            },
            "weight": Object {
              "max": 10000,
              "min": 500,
              "step": 1,
              "type": "number",
            },
            "year": Object {
              "max": 99,
              "min": 0,
              "step": 1,
              "type": "number",
            },
          },
          "onCancel": [Function],
          "onChangeField": [Function],
          "onSubmit": [Function],
          "validateField": [Function],
        },
        "refs": Object {},
        "state": Object {
          "car": Object {
            "acceleration": 12,
            "id": 1,
            "name": "Audi",
          },
        },
        "submit": [Function],
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <CarFormPresentation
          car={
                    Object {
                              "acceleration": 12,
                              "id": 1,
                              "name": "Audi",
                            }
          }
          errors={Object {}}
          model={
                    Object {
                              "acceleration": Object {
                                "max": 100,
                                "min": 1,
                                "step": 0.1,
                                "type": "number",
                              },
                              "cylinders": Object {
                                "max": 16,
                                "min": 2,
                                "step": 1,
                                "type": "number",
                              },
                              "displacement": Object {
                                "max": 1000,
                                "min": 1,
                                "step": 1,
                                "type": "number",
                              },
                              "horsepower": Object {
                                "max": 1000,
                                "min": 40,
                                "step": 1,
                                "type": "number",
                              },
                              "mpg": Object {
                                "max": 100,
                                "min": 0,
                                "step": 1,
                                "type": "number",
                              },
                              "name": Object {
                                "type": "string",
                              },
                              "weight": Object {
                                "max": 10000,
                                "min": 500,
                                "step": 1,
                                "type": "number",
                              },
                              "year": Object {
                                "max": 99,
                                "min": 0,
                                "step": 1,
                                "type": "number",
                              },
                            }
          }
          onCancel={[Function]}
          onChangeField={[Function]}
          onSubmit={[Function]}
/>,
        "_debugID": 2,
        "_renderedOutput": <CarFormPresentation
          car={
                    Object {
                              "acceleration": 12,
                              "id": 1,
                              "name": "Audi",
                            }
          }
          errors={Object {}}
          model={
                    Object {
                              "acceleration": Object {
                                "max": 100,
                                "min": 1,
                                "step": 0.1,
                                "type": "number",
                              },
                              "cylinders": Object {
                                "max": 16,
                                "min": 2,
                                "step": 1,
                                "type": "number",
                              },
                              "displacement": Object {
                                "max": 1000,
                                "min": 1,
                                "step": 1,
                                "type": "number",
                              },
                              "horsepower": Object {
                                "max": 1000,
                                "min": 40,
                                "step": 1,
                                "type": "number",
                              },
                              "mpg": Object {
                                "max": 100,
                                "min": 0,
                                "step": 1,
                                "type": "number",
                              },
                              "name": Object {
                                "type": "string",
                              },
                              "weight": Object {
                                "max": 10000,
                                "min": 500,
                                "step": 1,
                                "type": "number",
                              },
                              "year": Object {
                                "max": 99,
                                "min": 0,
                                "step": 1,
                                "type": "number",
                              },
                            }
          }
          onCancel={[Function]}
          onChangeField={[Function]}
          onSubmit={[Function]}
/>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <CarForm
    car={
        Object {
            "acceleration": 12,
            "id": 1,
            "name": "Audi",
          }
    }
    model={
        Object {
            "acceleration": Object {
              "max": 100,
              "min": 1,
              "step": 0.1,
              "type": "number",
            },
            "cylinders": Object {
              "max": 16,
              "min": 2,
              "step": 1,
              "type": "number",
            },
            "displacement": Object {
              "max": 1000,
              "min": 1,
              "step": 1,
              "type": "number",
            },
            "horsepower": Object {
              "max": 1000,
              "min": 40,
              "step": 1,
              "type": "number",
            },
            "mpg": Object {
              "max": 100,
              "min": 0,
              "step": 1,
              "type": "number",
            },
            "name": Object {
              "type": "string",
            },
            "weight": Object {
              "max": 10000,
              "min": 500,
              "step": 1,
              "type": "number",
            },
            "year": Object {
              "max": 99,
              "min": 0,
              "step": 1,
              "type": "number",
            },
          }
    }
    onCancel={[Function]}
    onChangeField={[Function]}
    onSubmit={[Function]}
    validateField={[Function]}
/>,
}
`;
